File: swift/src/Socket/ProxySocket/SOCKS5ProxySocket.swift
import Foundation

public class SOCKS5ProxySocket: ProxySocket {
    enum SOCKS5ProxyReadStatus: CustomStringConvertible {
        case invalid,
        readingVersionIdentifierAndNumberOfMethods,
        readingMethods,
        readingConnectHeader,
        readingIPv4Address,
        readingDomainLength,
        readingDomain,
        readingIPv6Address,
        readingPort,
        forwarding,
        stopped

        var description: String {
            switch self {
            case .invalid:
                return "invalid"
            case .readingVersionIdentifierAndNumberOfMethods:
                return "reading version and methods"
            case .readingMethods:
                return "reading methods"
            case .readingConnectHeader:
                return "reading connect header"
            case .readingIPv4Address:
                return "IPv4 address"
            case .readingDomainLength:
                return "domain length"
            case .readingDomain:
                return "domain"
            case .readingIPv6Address:
                return "IPv6 address"
            case .readingPort:
                return "reading port"
            case .forwarding:
                return "forwarding"
            case .stopped:
                return "stopped"
            }
        }
    }

    enum SOCKS5ProxyWriteStatus: CustomStringConvertible {
        case invalid,
        sendingResponse,
        forwarding,
        stopped

        var description: String {
            switch self {
            case .invalid:
                return "invalid"
            case .sendingResponse:
                return "sending response"
            case .forwarding:
                return "forwarding"
            case .stopped:
                return "stopped"
            }
        }
    }
    /// The remote host to connect to.
    public var destinationHost: String!

    /// The remote port to connect to.
    public var destinationPort: Int!

    public var outboundInterfaceType: NetworkInterfaceType = .default
    
    /// The address the SOCKS5 server is listening on. Used as BND.ADDR in UDP ASSOCIATE replies.
    public var serverAddress: IPAddress?
    
    // Holds the relay server if this is a UDP ASSOCIATE request
    private var udpRelayServer: SOCKS5UDPRelayServer?
    
    // Command type requested (1 = CONNECT, 3 = UDP ASSOCIATE)
    private var requestedCommand: UInt8 = 1

    private var readStatus: SOCKS5ProxyReadStatus = .invalid
    private var writeStatus: SOCKS5ProxyWriteStatus = .invalid

    public var readStatusDescription: String {
        return readStatus.description
    }

    public var writeStatusDescription: String {
        return writeStatus.description
    }

    /**
     Begin reading and processing data from the socket.
     */
    override public func openSocket() {
        super.openSocket()

        guard !isCancelled else {
            return
        }

        readStatus = .readingVersionIdentifierAndNumberOfMethods
        socket.readDataTo(length: 2)
    }

    // swiftlint:disable function_body_length
    // swiftlint:disable cyclomatic_complexity
    /**
     The socket did read some data.
     
     - parameter data:    The data read from the socket.
     - parameter from:    The socket where the data is read from.
     */
    override public func didRead(data: Data, from: RawTCPSocketProtocol) {
        super.didRead(data: data, from: from)

        switch readStatus {
        case .forwarding:
            delegate?.didRead(data: data, from: self)
        case .readingVersionIdentifierAndNumberOfMethods:
            data.withUnsafeBytes { (pointer: UnsafePointer<UInt8>) in
                guard pointer.pointee == 5 else {
                    // TODO: notify observer
                    self.disconnect()
                    return
                }

                guard pointer.successor().pointee > 0 else {
                    // TODO: notify observer
                    self.disconnect()
                    return
                }

                self.readStatus = .readingMethods
                self.socket.readDataTo(length: Int(pointer.successor().pointee))
            }
        case .readingMethods:
            // TODO: check for 0x00 in read data

            let response = Data(bytes: [0x05, 0x00])
            // we would not be able to read anything before the data is written out, so no need to handle the dataWrote event.
            write(data: response)
            readStatus = .readingConnectHeader
            socket.readDataTo(length: 4)
        case .readingConnectHeader:
            data.withUnsafeBytes { (pointer: UnsafePointer<UInt8>) in
                let cmd = pointer.successor().pointee
                guard pointer.pointee == 5 && (cmd == 1 || cmd == 3) else {
                    // TODO: notify observer
                    self.disconnect()
                    return
                }
                
                self.requestedCommand = cmd
                
                switch pointer.advanced(by: 3).pointee {
                case 1:
                    readStatus = .readingIPv4Address
                    socket.readDataTo(length: 4)
                case 3:
                    readStatus = .readingDomainLength
                    socket.readDataTo(length: 1)
                case 4:
                    readStatus = .readingIPv6Address
                    socket.readDataTo(length: 16)
                default:
                    break
                }
            }
        case .readingIPv4Address:
            var address = Data(count: Int(INET_ADDRSTRLEN))
            _ = data.withUnsafeRawPointer { data_ptr in
                address.withUnsafeMutableBytes { addr_ptr in
                    inet_ntop(AF_INET, data_ptr, addr_ptr, socklen_t(INET_ADDRSTRLEN))
                }
            }

            address.withUnsafeBytes {
                destinationHost = String(cString: $0, encoding: .utf8)
            }

            readStatus = .readingPort
            socket.readDataTo(length: 2)
        case .readingIPv6Address:
            var address = Data(count: Int(INET6_ADDRSTRLEN))
            _ = data.withUnsafeRawPointer { data_ptr in
                address.withUnsafeMutableBytes { addr_ptr in
                    inet_ntop(AF_INET6, data_ptr, addr_ptr, socklen_t(INET6_ADDRSTRLEN))
                }
            }

            address.withUnsafeBytes {
                destinationHost = String(cString: $0, encoding: .utf8)
            }

            readStatus = .readingPort
            socket.readDataTo(length: 2)
        case .readingDomainLength:
            data.withUnsafeRawPointer {
                readStatus = .readingDomain
                socket.readDataTo(length: Int($0.load(as: UInt8.self)))
            }
        case .readingDomain:
            destinationHost = String(data: data, encoding: .utf8)
            readStatus = .readingPort
            socket.readDataTo(length: 2)
        case .readingPort:
            data.withUnsafeRawPointer {
                destinationPort = Int($0.load(as: UInt16.self).bigEndian)
            }

            if requestedCommand == 3 {
                // UDP ASSOCIATE
                startUDPRelayServer()
            } else {
                // TCP CONNECT
                readStatus = .forwarding
                session = ConnectSession(host: destinationHost, port: destinationPort)
                session?.interfaceType = outboundInterfaceType
                observer?.signal(.receivedRequest(session!, on: self))
                delegate?.didReceive(session: session!, from: self)
            }
        default:
            return
        }
    }
    
    private func startUDPRelayServer() {
        guard let destHost = destinationHost, let destPort = destinationPort else {
            disconnect()
            return
        }
        
        // Use 0.0.0.0 as source hint for relay if domain was provided, though relay handles Any.
        let parsedAddr = IPAddress(fromString: destHost) ?? IPAddress(fromString: "0.0.0.0")!
        
        let relay = SOCKS5UDPRelayServer(
            expectedClientAddress: parsedAddr,
            expectedClientPort: Port(port: UInt16(destPort)),
            socks5ProxySocket: self,
            outboundInterfaceType: outboundInterfaceType,
            bindAddress: serverAddress
        )
        
        if relay.start() {
            self.udpRelayServer = relay
            
            // Reply with success and bound address/port
            var responseBytes = [UInt8](repeating: 0, count: 10)
            responseBytes[0...3] = [0x05, 0x00, 0x00, 0x01]
            // Per RFC 1928, BND.ADDR should be the address the client can reach.
            // Use the SOCKS5 server's listening address.
            let replyIp = serverAddress?.presentation ?? "0.0.0.0"
            let boundIpBytes = replyIp.components(separatedBy: ".").compactMap { UInt8($0) }
            for (i, byte) in boundIpBytes.enumerated() {
                if i < 4 { responseBytes[4 + i] = byte }
            }
            let port = relay.boundPort?.value ?? 0
            responseBytes[8] = UInt8(port >> 8)
            responseBytes[9] = UInt8(port & 0xFF)
            
            let responseData = Data(bytes: responseBytes)
            
            // For UDP ASSOCIATE, the TCP connection stays open just to keep the UDP relay alive.
            // We just enter a dummy forwarding state or a "wait for close" state.
            writeStatus = .forwarding
            readStatus = .forwarding 
            write(data: responseData)
        } else {
            // Send failure (General SOCKS server failure)
            var responseBytes = [UInt8](repeating: 0, count: 10)
            responseBytes[0...3] = [0x05, 0x01, 0x00, 0x01]
            let responseData = Data(bytes: responseBytes)
            writeStatus = .stopped
            write(data: responseData)
            disconnect()
        }
    }

    /**
     The socket did send some data.
     
     - parameter data:    The data which have been sent to remote (acknowledged). Note this may not be available since the data may be released to save memory.
     - parameter from:    The socket where the data is sent out.
     */
    override public func didWrite(data: Data?, by: RawTCPSocketProtocol) {
        super.didWrite(data: data, by: by)

        switch writeStatus {
        case .forwarding:
            delegate?.didWrite(data: data, by: self)
        case .sendingResponse:
            writeStatus = .forwarding
            observer?.signal(.readyForForward(self))
            delegate?.didBecomeReadyToForwardWith(socket: self)
        default:
            return
        }
    }

    /**
     Response to the `AdapterSocket` on the other side of the `Tunnel` which has succefully connected to the remote server.
     
     - parameter adapter: The `AdapterSocket`.
     */
    override public func respondTo(adapter: AdapterSocket) {
        super.respondTo(adapter: adapter)

        guard !isCancelled else {
            return
        }

        var responseBytes = [UInt8](repeating: 0, count: 10)
        responseBytes[0...3] = [0x05, 0x00, 0x00, 0x01]
        let responseData = Data(bytes: responseBytes)

        writeStatus = .sendingResponse
        write(data: responseData)
    }
}

File: kotlin/src/main/kotlin/Socket/ProxySocket/SOCKS5ProxySocket.kt
package nekit.Socket.ProxySocket

import nekit.Messages.ConnectSession
import nekit.ProxyServer.SOCKS5UDPRelayServer
import nekit.RawSocket.protocol.RawTCPSocketProtocol
import nekit.Socket.AdapterSocket.AdapterSocket
import nekit.Socket.SocketStatus
import nekit.Utils.IPAddress
import nekit.Utils.Port
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch
import org.slf4j.LoggerFactory
import java.io.IOException
import nekit.Config.NetworkInterfaceType

class SOCKS5ProxySocket(
    rawSocket: RawTCPSocketProtocol
) : ProxySocket(rawSocket) {

    override var session: ConnectSession? = null
        public set

    var outboundInterfaceType: NetworkInterfaceType = NetworkInterfaceType.DEFAULT

    /// The address the SOCKS5 server is listening on. Used as BND.ADDR in UDP ASSOCIATE replies.
    var serverAddress: IPAddress? = null

    private val logger = LoggerFactory.getLogger(SOCKS5ProxySocket::class.java)
    private var state = State.INITIAL
    private val socketScope = CoroutineScope(SupervisorJob() + Dispatchers.Default)
    
    // For storing the relay server if it's a UDP associate connection
    private var udpRelayServer: SOCKS5UDPRelayServer? = null
    // To identify if the current connection is for UDP
    private var isUdpAssociate = false

    private enum class State {
        INITIAL,
        GREETING,
        READING_METHODS,
        CONNECTING,
        READING_IPV4,
        READING_IPV6,
        READING_DOMAIN_LENGTH,
        READING_DOMAIN,
        READING_PORT,
        SENDING_RESPONSE,
        FORWARDING
    }

    override fun openSocket() {
        super.openSocket()
        state = State.GREETING
        // Start by reading exactly 2 bytes for version and number of methods
        rawSocket.readDataTo(2)
    }

    override fun didRead(data: ByteArray, from: RawTCPSocketProtocol) {
        when (state) {
            State.GREETING -> {
                handleGreeting(data)
            }
            State.READING_METHODS -> {
                handleMethods(data)
            }
            State.CONNECTING -> {
                handleConnect(data)
            }
            State.READING_IPV4 -> {
                handleIPv4Address(data)
            }
            State.READING_IPV6 -> {
                handleIPv6Address(data)
            }
            State.READING_DOMAIN_LENGTH -> {
                handleDomainLength(data)
            }
            State.READING_DOMAIN -> {
                handleDomain(data)
            }
            State.READING_PORT -> {
                handlePort(data)
            }
            State.FORWARDING -> {
                delegate?.get()?.didRead(data, this)
            }
            else -> return
        }
    }

    override fun didWrite(data: ByteArray?, by: RawTCPSocketProtocol) {
        super.didWrite(data, by)
        when (state) {
            State.SENDING_RESPONSE -> {
                state = State.FORWARDING
                _status = SocketStatus.ESTABLISHED
                delegate?.get()?.didBecomeReadyToForward(this)
            }
            State.FORWARDING -> {
                delegate?.get()?.didWrite(data, this)
            }
            else -> return
        }
    }

    private fun handleGreeting(data: ByteArray) {
        // Should receive exactly 2 bytes: version and number of methods
        if (data.size < 2) {
            forceDisconnect(IOException("Invalid SOCKS5 greeting: insufficient data"))
            return
        }
        
        val version = data[0]
        val nMethods = data[1].toInt() and 0xFF
        
        if (version != 0x05.toByte()) {
            forceDisconnect(IOException("Invalid SOCKS5 version: $version"))
            return
        }
        
        if (nMethods <= 0) {
            forceDisconnect(IOException("Invalid number of methods: $nMethods"))
            return
        }
        
        // Read the methods
        rawSocket.readDataTo(nMethods) 
        state = State.READING_METHODS
    }
    
    private fun handleMethods(data: ByteArray) {
        // TODO: check for 0x00 in read data
        val response = byteArrayOf(0x05, 0x00) // NO AUTH
        write(response)
        // After sending response, read connect header (4 bytes)
        rawSocket.readDataTo(4)
        state = State.CONNECTING
    }

    private fun handleConnect(data: ByteArray) {
        // Should receive exactly 4 bytes: VER, CMD, RSV, ATYP
        if (data.size < 4) {
            forceDisconnect(IOException("Invalid SOCKS5 connect header: insufficient data"))
            return
        }
        
        val version = data[0]
        val cmd = data[1]
        // Skip RSV (data[2])
        val atyp = data[3]
        
        if (version != 0x05.toByte()) {
            forceDisconnect(IOException("Invalid SOCKS5 connect request: version=$version"))
            return
        }
        
        if (cmd == 0x01.toByte()) {
            isUdpAssociate = false
        } else if (cmd == 0x03.toByte()) {
            isUdpAssociate = true
        } else {
            // We only support CONNECT (0x01) and UDP ASSOCIATE (0x03)
            forceDisconnect(IOException("Unsupported SOCKS5 command: $cmd"))
            return
        }
        
        // Based on address type, read the appropriate amount of data
        when (atyp) {
            0x01.toByte() -> { // IPv4 - read 4 bytes for IP
                state = State.READING_IPV4
                rawSocket.readDataTo(4) 
            }
            0x03.toByte() -> { // Domain - read 1 byte for length first
                state = State.READING_DOMAIN_LENGTH
                rawSocket.readDataTo(1) 
            }
            0x04.toByte() -> { // IPv6 - read 16 bytes for IP
                state = State.READING_IPV6
                rawSocket.readDataTo(16) 
            }
            else -> {
                forceDisconnect(IOException("Unsupported address type in SOCKS5 request: $atyp"))
            }
        }
    }
    
    private var destinationHost: String? = null
    private var targetPort: Int? = null
    
    private fun handleIPv4Address(data: ByteArray) {
        if (data.size < 4) {
            forceDisconnect(IOException("Invalid IPv4 address data"))
            return
        }
        destinationHost = data.joinToString(".") { (it.toInt() and 0xFF).toString() }
        state = State.READING_PORT
        rawSocket.readDataTo(2)
    }
    
    private fun handleIPv6Address(data: ByteArray) {
        if (data.size < 16) {
            forceDisconnect(IOException("Invalid IPv6 address data"))
            return
        }
        destinationHost = data.asList().chunked(2).joinToString(":") {
            String.format("%02x%02x", it[0], it[1])
        }
        state = State.READING_PORT
        rawSocket.readDataTo(2)
    }
    
    private fun handleDomainLength(data: ByteArray) {
        if (data.isEmpty()) {
            forceDisconnect(IOException("Invalid domain length data"))
            return
        }
        val domainLength = data[0].toInt() and 0xFF
        state = State.READING_DOMAIN
        rawSocket.readDataTo(domainLength)
    }
    
    private fun handleDomain(data: ByteArray) {
        destinationHost = String(data)
        state = State.READING_PORT
        rawSocket.readDataTo(2)
    }
    
    private fun handlePort(data: ByteArray) {
        if (data.size < 2) {
            forceDisconnect(IOException("Invalid port data"))
            return
        }
        targetPort = ((data[0].toInt() and 0xFF) shl 8) or (data[1].toInt() and 0xFF)
        
        val host = destinationHost
        val port = targetPort
        if (host != null && port != null) {
            val requestSession = ConnectSession(host = host, port = port, interfaceType = outboundInterfaceType)
            this.session = requestSession
            
            if (isUdpAssociate) {
                // Determine client's assumed IP and Port from the connection itself
                // (Using the TCP source IP as best-effort for UDP source if the client sent 0.0.0.0)
                val clientIP = rawSocket.sourceIPAddress ?: IPAddress.parse("0.0.0.0")!!
                var clientPortParam: Port = Port(port)
                // If the client sent 0.0.0.0:0, it means any IP and Port might be used. 
                // We'll trust our SOCKS5UDPRelayServer's default acceptance logic.
                
                startUdpRelay(clientIP, clientPortParam)
            } else {
                // 不要立即进入FORWARDING状态，等待respondTo被调用
                delegate?.get()?.didReceive(requestSession, this)
            }
        } else {
            forceDisconnect(IOException("Missing host or port information"))
        }
    }
    
    private fun startUdpRelay(clientIP: IPAddress, clientPort: Port) {
        socketScope.launch {
            val relayServer = SOCKS5UDPRelayServer(clientIP, clientPort, this@SOCKS5ProxySocket, outboundInterfaceType, bindAddress = serverAddress)
            val success = relayServer.start()
            
            if (success) {
                udpRelayServer = relayServer
                val boundIPStr = serverAddress?.presentation ?: "0.0.0.0"
                val boundPortInt = relayServer.boundPort?.hostOrderValue ?: 0
                
                // Reply to the client with the IP and Port they should send UDP datagrams to
                sendUdpAssociateSuccessResponse(boundIPStr, boundPortInt)
            } else {
                logger.error("Failed to start UDP relay server.")
                // Send general SOCKS server failure reply
                write(byteArrayOf(0x05, 0x01, 0x00, 0x01, 0, 0, 0, 0, 0, 0))
                forceDisconnect(IOException("Failed to start UDP Relay"))
            }
        }
    }
    
    private fun sendUdpAssociateSuccessResponse(boundIP: String, boundPort: Int) {
        val ipParts = boundIP.split(".")
        if (ipParts.size == 4) {
            // IPv4
            val response = ByteArray(10)
            response[0] = 0x05 // VER
            response[1] = 0x00 // REP (Success)
            response[2] = 0x00 // RSV
            response[3] = 0x01 // ATYP (IPv4)
            for (i in 0..3) {
                response[4 + i] = ipParts[i].toInt().toByte()
            }
            response[8] = ((boundPort shr 8) and 0xFF).toByte()
            response[9] = (boundPort and 0xFF).toByte()
            
            state = State.SENDING_RESPONSE
            write(response)
        } else {
            // Simplistic fallback for IPv6
            val response = ByteArray(10)
            response[0] = 0x05 // VER
            response[1] = 0x00 // REP (Success)
            response[2] = 0x00 // RSV
            response[3] = 0x01 // ATYP (IPv4)
            // returning 0.0.0.0 for IP
            response[4] = 0; response[5] = 0; response[6] = 0; response[7] = 0;
            response[8] = ((boundPort shr 8) and 0xFF).toByte()
            response[9] = (boundPort and 0xFF).toByte()
            state = State.SENDING_RESPONSE
            write(response)
        }
    }

    override fun respondTo(adapter: AdapterSocket) {
        // Do NOT call super.respondTo() - that would fire didBecomeReadyToForward immediately.
        // Instead, we manually signal the observer and send the SOCKS5 success reply.
        // didBecomeReadyToForward will be fired exactly once in didWrite when
        // state transitions from SENDING_RESPONSE → FORWARDING.
        if (isCancelled) {
            logger.warn("respondTo called on a cancelled socket for session: {}", session)
            return
        }
        logger.info("respondTo called with adapter {} for session {}.", adapter, session)
        observer?.signal(nekit.Event.Event.ProxySocketEvent.AskedToResponseTo(adapter, this))
        // SOCKS5 success reply: VER | REP | RSV | ATYP | BND.ADDR | BND.PORT
        val response = byteArrayOf(0x05, 0x00, 0x00, 0x01, 0, 0, 0, 0, 0, 0)
        state = State.SENDING_RESPONSE
        write(response)
    }
    override fun forceDisconnect(becauseOf: Throwable?) {
        socketScope.cancel()
        udpRelayServer?.stop()
        udpRelayServer = null
        super.forceDisconnect(becauseOf)
    }
}
